### PR TITLE
Added a timer for escaping(currently 5 sec)

### LINKED ISSUE #861: Make escaping from engulfing population effect more accurate
It's possible that a species gets the population buff from not actually escaping something that could kill it. The escaping from engulfing effect should be made only apply when a cell escapes something that could actually kill it.
https://community.revolutionarygamesstudio.com/t/0-4-2-feedback-thread/2143?u=hhyyrylainen

I'm also marking this issue as review as I didn't personally verify this.

### PR DESCRIPTION
**Pull request by Savalige**
fixes #861

### CODE REVIEW

**Comment by hhyyrylainen** [comment]
add to whitelist

File: scripts/microbe_stage/microbe.as [hunk 1]
```diff
@@ -433,14 +434,17 @@ class MicrobeSystem : ScriptSystem{
             microbeComponent.movementFactor =  microbeComponent.movementFactor/ENGULFING_MOVEMENT_DIVISION;
         }
 
+    microbeComponent.escapeInterval = microbeComponent.escapeInterval + logicTime;
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe.as]: The indent level doesn't match up here and a couple of other spots.

Comment (Savalige) [reply scripts/microbe_stage/microbe.as]: Will do but then the format script hates it

Comment (Savalige) [reply scripts/microbe_stage/microbe.as]: Ah think I have found the problem

File: scripts/microbe_stage/microbe.as [hunk 2]
```diff
@@ -433,14 +434,17 @@ class MicrobeSystem : ScriptSystem{
             microbeComponent.movementFactor =  microbeComponent.movementFactor/ENGULFING_MOVEMENT_DIVISION;
         }
 
+        microbeComponent.escapeInterval = microbeComponent.escapeInterval + logicTime;
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe.as]: Should be written as:
`microbeComponent.escapeInterval += logicTime;`

File: scripts/microbe_stage/microbe.as [hunk 3]
```diff
@@ -433,14 +434,17 @@ class MicrobeSystem : ScriptSystem{
...
         if(microbeComponent.isBeingEngulfed){
             microbeComponent.movementFactor =  microbeComponent.movementFactor/ENGULFED_MOVEMENT_DIVISION;
             //LOG_INFO("doing engulf damage");
             MicrobeOperations::damage(world,microbeEntity,ENGULF_DAMAGE/logicTime,
                 "isBeingEngulfed - Microbe.update()s");
             microbeComponent.wasBeingEngulfed = true;
+            microbeComponent.escapeInterval = 0;
             // Else If we were but are no longer, being engulfed
-        } else if(microbeComponent.wasBeingEngulfed && !microbeComponent.isBeingEngulfed){
+        } else if(microbeComponent.wasBeingEngulfed && !microbeComponent.isBeingEngulfed && microbeComponent.escapeInterval > CREATURE_ESCAPE_INTERVAL){
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe.as]: I feel like this is the wrong place to prevent the engulfing from stopping. This interval check being here prevents `removeEngulfedEffect` from being called.

**Comment by Savalige** [comment]
Yeah but since that is only called once the timer would still be 0

**Comment by hhyyrylainen** [comment]
I thought that it would work like this:
When triggering the population effect check that `escapeEngulfTimeOut <= 0`
and then add the population effect, after adding the effect set `escapeEngulfTimeOut = 5000`

Now then put in the general cell handling code to decrement escapeEngulfTimeOut if it is > 0.

**Comment by Savalige** [comment]
Ah I was thinking more in the way of that you have to not be engulfed for a while to escape

**Comment by hhyyrylainen** [comment]
I suppose that is also a good approach, but I think that even then the check that you haven't been engulfed for n number of seconds after escaping needs to be in the general cell tick and not in the same place as handling engulfing stopping.

**Comment by Savalige** [comment]
Are you talking about just not having it as an else if? Not really sure what you mean

**Comment by hhyyrylainen** [comment]
I'm talking about code like this:
```
            if (!microbeComponent.isPlayerMicrobe &&
                microbeComponent.species.name != playerSpecies.name)
            {
                 hasEscaped = true;
                 timeSinceEscape = 0;
            }
```

Then somewhere else there is code to increase that counter and trigger the action:
```
if(hasEscaped){
    timeSinceEscape += logicTime;
       if(timeSinceEscape >= 5000){
                hasEscaped = false;
                timeSinceEscape = 0;
                auto species = MicrobeOperations::getSpecies(world,
                    microbeComponent.species.name);
                if(species !is null)
                    MicrobeOperations::alterSpeciesPopulation(species,
                        CREATURE_ESCAPE_POPULATION_GAIN, "escape engulfing");
```

With code like that + the two new properties in the microbe component, it can be tracked that a cell needs to avoid engulfing for some time in order to get the population.

**Comment by Savalige** [comment]
If i understood you correctly this should be correct but i did change hasEscaped since I think that it should be true if you have escaped

**Comment by Savalige** [comment]
NO I HAVE MISSED SOMETHING

**Comment by Savalige** [comment]
Was thinking that it should be this way to have hasEscaped is true when it gets the population added as if saying that you have not technically escaped until this happens.
What condition are you talking about specifically.

**Comment by hhyyrylainen** [comment]
To me it seems more intuitive to have it so that when the engulfing stops the cell has technically escaped, but it will only be given the population increase if it manages to stay away from engulfers for 5 seconds.

**Comment by Savalige** [comment]
Ah I see what you mean can change that

File: scripts/microbe_stage/microbe.as [hunk 4]
```diff
@@ -160,6 +160,7 @@ class MicrobeComponent : ScriptComponent{
     float maxBandwidth = 10.0 * BANDWIDTH_PER_ORGANELLE; // wtf is a bandwidth anyway?
     float remainingBandwidth = 0.0;
     uint compoundCollectionTimer = EXCESS_COMPOUND_COLLECTION_INTERVAL;
+    uint escapeInterval = CREATURE_ESCAPE_INTERVAL;
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe.as]: Shouldn't this be initialized to zero?

File: scripts/microbe_stage/microbe.as [hunk 5]
```diff
@@ -438,26 +440,34 @@ class MicrobeSystem : ScriptSystem{
             //LOG_INFO("doing engulf damage");
             MicrobeOperations::damage(world,microbeEntity,ENGULF_DAMAGE/logicTime,
                 "isBeingEngulfed - Microbe.update()s");
-            microbeComponent.wasBeingEngulfed = true;
+            microbeComponent.wasBeingEngulfed = false;
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe.as]: Why is this change here? This PR shouldn't change the engulf mechanics.

File: scripts/microbe_stage/microbe.as [hunk 6]
```diff
@@ -438,26 +440,34 @@ class MicrobeSystem : ScriptSystem{
...
                 microbeComponent.species.name != playerSpecies.name)
             {
+                 microbeComponent.hasEscaped = true;
+                 microbeComponent.escapeInterval = 0;
+            }
+
+            MicrobeOperations::removeEngulfedEffect(world, microbeEntity);
+        }
+
+        // Still considered to be chased for CREATURE_ESCAPE_INTERVAL milisecunds
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe.as]: typo here

File: scripts/microbe_stage/microbe.as [hunk 7]
```diff
@@ -438,26 +440,34 @@ class MicrobeSystem : ScriptSystem{
             //LOG_INFO("doing engulf damage");
             MicrobeOperations::damage(world,microbeEntity,ENGULF_DAMAGE/logicTime,
                 "isBeingEngulfed - Microbe.update()s");
-            microbeComponent.wasBeingEngulfed = true;
+            microbeComponent.wasBeingEngulfed = false;
             // Else If we were but are no longer, being engulfed
         } else if(microbeComponent.wasBeingEngulfed && !microbeComponent.isBeingEngulfed){
-
```
Comment (hhyyrylainen) [inline scripts/microbe_stage/microbe.as]: This and the next blank line should be kept.

File: scripts/microbe_stage/microbe.as (hunk 4 above)
Comment (hhyyrylainen) [reply scripts/microbe_stage/microbe.as]: This is still here.

Comment (Savalige) [reply scripts/microbe_stage/microbe.as]: Ah yes missed that

**Comment by hhyyrylainen** [comment]
I didn't test how well this works in game. I'm going to assume that you did enough testing.
